### ISSUE TITLE
Tests fail until Accessibility Inspector is run

### ISSUE THREAD

**Issue by davbeck** [kif-framework/KIF#944]
We have a KIF test suite that runs just fine on some Macs, but has failed on our Xcode Server and some team members Macs for a while. The tests would start the simulator but timeout on the first accessibility action. After several days of investigating I've found that if I open the Accessibility Inspector and switch it to the simulator, even if our app isn't open, will fix the issues. Resetting the simulator resets the issue.

Has anyone else seen this kind of issue? Is there anyway to automate a fix?

**Comment by phatmann** [comment]
KIF has some special code that is supposed to open the Accessibility Inspector for just this reason. Not sure why it is not working for you. Which version of iOS and Xcode are you using?

**Comment by davbeck** [comment]
Xcode 8.2 and iOS 10.2 simulator. Is it related to the switch to Xcode using a Mac app for accessibility instead of the floating iOS one?

**Comment by LMojica** [comment]
I am also experiencing this problem. It doesn't happen every time, but maybe about 5% of the time. I launch a test, and it fails with this error. 

`2017-02-13 16:29:23.736 MyApp[92242:9744281] UIAutomation can't work because Accessibility API is not enabled.`

**Comment by justinseanmartin** [comment]
@davbeck - Are you using the latest version of KIF? I'd recommend setting a breakpoint in `KIFAccessibilityEnabler.m` and make sure that `_AXSSetAutomationEnabled` is being called successfully. This is the code path that I'm hitting with Xcode 8.2.1 & iOS 10.2 on the latest KIF sources. 

If things are working correctly, running `po [UIView printViewHierarchy]` in lldb should give the element hierarchy with labels & traits.

One thing I just noticed is that in KIF, `KIFEnableAccessibility()` is being called in `+[KIFTestCase setUp]`. I've seen instances in the past where `+setUp` wouldn't get called (e.g. running a filtered set of tests from the Xcode diamonds). I can't reproduce this at the moment, but in our tests we explicitly call `KIFAccessibilityEnabler()` in a `dispatch_once` block from our base classes `-setUp` method before using KIF to find any UI elements. This has been working consistently for us. If this helps, maybe we can consider adding a different hook point for calling it.

**Comment by davbeck** [comment]
This was the key right here. Apparently our base test class wasn't inheriting from `KIFTestCase`.

**Comment by anshu10165** [comment]
My Test-Class in inheriting from a class `MYKIFTestCase` which again inherits from `KIFTestCase`.Using `KIFEnableAccessibility()` in `setUp` of `MYKIFTestCase` didn't work for me.Error `failed: caught **kUIAExeceptionNoAccessibility", "UIAutomation can't work because Accessibility API is not enabled.**`. Could someone please help here?

**Comment by mikelupo** [comment]
When you created your project did you create a UI test project or a unit test project?
Should be unit test project.

**Comment by anshu10165** [comment]
Hey @mikelupo , thanks for the info. Any way I can check if my project is under UI or Unit testing target ? I recently joined this org, and still adapting myself to iOS and KIF.Any help would be much appreciated here. The target `KIFTests` created long time back here which is dedicated to contain UI tests only, but I cannot figure out under which template(UI or Unit testing bundle) this was made.

**Comment by justinseanmartin** [comment]
I'm actually not sure where in the Xcode project settings UI it is represented that a target is a UI test versus a unit test. You can look in the `*.xcodeproj/project.pbxproj` in a text editor and see if the target that you are trying to run the tests with has `productType = "com.apple.product-type.bundle.ui-testing";`.

**Comment by anshu10165** [comment]
Thanks @justinseanmartin, I checked and the `pbxproj` has my `KIFTests` target categorized under `productType = "com.apple.product-type.bundle.unit-testing"` only. Not sure what went wrong in my case, still trying to figure out the `UIAutomation wont work because Accessibility Api is not enabled` error and no luck so far .

**Comment by mikelupo** [comment]
Actual device or sim?

**Comment by anshu10165** [comment]
@mikelupo : errors on iOS Simulators!

**Comment by ajfigueroa** [comment]
I'm seeing this issue as well but strange enough it only happens on CI and not always.

**Comment by justinseanmartin** [comment]
@ajfigueroa - You're seeing the same error of "UIAutomation wont work because Accessibility Api is not enabled"?

Just to double check, you're using the latest published version of KIF (3.6.0), correct? And your tests all subclass from `KIFTestCase` as called out above? I don't know of a reason why it wouldn't work consistently, we've had no issues with this being unreliable.

**Comment by ajfigueroa** [comment]
@justinseanmartin Yup, I am using a fork that's up to date with upstream and the failing test inherits from `KIFTestCase`.

This seems to be CI (Bitrise in this case) specific as changing devices no longer shows this issue. I'll do some more investigation there.

The problem never occurs locally though thankfully 👍

**Comment by maryamaffinityclick** [comment]
Im getting same error `UIAutomation can't work because Accessibility API is not enabled. failed: caught "kUIAExeceptionNoAccessibility", "UIAutomation can't work because Accessibility API is not enabled." `
It happens sometimes, I guess when first time opening the simulator on running my test and also I m running more than one simulator to run the tests !
sometimes solved by running the Accessibility Inspector and choose the simulators once and run the simulators and make them ready to run the test ...

**Comment by albertomota-codes** [comment]
Same error. I need to restart the simulator. Every time i restart the simulator, the test works again. Delete the app is not enough! :S
